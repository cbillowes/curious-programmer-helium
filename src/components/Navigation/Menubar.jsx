import React, { Component } from "react"
import Link from "gatsby-link"
import navigation from "../../../data/Navigation"
import "./Menubar.scss"

class Menubar extends Component {
  constructor(props) {
    super(props)
  }

  getRootUrl = () => {
    if (typeof window !== "undefined" && window.location) {
      let url = window.location.pathname
      if (url !== "/") {
        let paths = url.split("/")
        if (paths[1] === "blog") {
          return "/"
        } else {
          return `/${paths[1]}`
        }
      }
      return url
    }
  }

  isSelected = (url) => {
    return this.getRootUrl() === url ? "selected" : ""
  }

  render() {
    return (
      <nav
        className={`menubar ${this.props.active ? "active" : ""}`}
        onClick={this.props.toggle.bind(this)}
      >
        <ul>
          {navigation.map((item) => {
            return (
              <li
                key={`menubar-${item.url}`}
                className={`menu-item ${this.isSelected(item.url)}`}
              >
                <Link to={item.url}>{item.name}</Link>
              </li>
            )
          })}
        </ul>
      </nav>
    )
  }
}

export default Menubar