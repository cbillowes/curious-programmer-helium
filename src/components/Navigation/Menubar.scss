@import "../../layout/variables.scss";

@mixin flyout {
    opacity: 0;
    -webkit-transform: translate3d(300px, 0, 0);
    -ms-transform: translate3d(300px, 0, 0);
    transform: translate3d(300px, 0, 0);
    -webkit-transition: -webkit-transform 0.5s ease, opacity 0.3s ease 0.7s;
    transition: transform 0.5s ease, opacity 0.3s ease 0.7s;
}

@mixin flyin {
    opacity: 1;
    -webkit-transition: -webkit-transform 0.3s ease, opacity 0s ease 0s;
    transition: transform 0.3s ease, opacity 0s ease 0s;
    -webkit-transform: translate3d(0, 0, 0);
    -ms-transform: translate3d(0, 0, 0);
    transform: translate3d(0, 0, 0);
}

.navigation {
    .menubar {
        background-color: $background-darker;
        position: fixed;
        width: $menubar-width;
        top: 4em;
        right: 0;
        height: 100vh;
        z-index: 999;
        text-shadow: none;

        @include flyout;

        &.active {
            @include flyin;
        }

        ul {
            padding-left: 0;
            margin: 0;

            li.menu-item {
                list-style: none;

                &.selected {
                    background-color: $blue;

                    a {
                        color: $white;
                    }
                }

                a {
                    display: block;
                    padding: 0.25em 2em;
                    font-family: $font-family-sans-serif;
                    color: $foreground;

                    .github {
                        margin-top: 0.5em;
                        margin-left: 0.5em;

                        path {
                            fill: $foreground;
                        }
                    }

                    &:hover {
                        background-color: $primary;
                        color: $background-lighter;

                        .github path {
                            fill: $background;
                        }
                    }
                }
            }
        }
    }
}
