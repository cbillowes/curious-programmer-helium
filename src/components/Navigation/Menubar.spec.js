import React from "react"
import renderer from "react-test-renderer"
import { mount } from "enzyme"
import Menubar from "./Menubar"

describe("a menu bar component", () => {

  it("should render", () => {
    let element = <Menubar toggle={() => { }} active="true" />
    const tree = renderer.create(element).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("should show the menubar when active", () => {
    const element = <Menubar toggle={() => { }} active={true} />
    const component = mount(element)
    expect(component.html()).toContain(`<nav class="menubar active">`)
  })

  it("should hide the menubar when it is not active", () => {
    const element = <Menubar toggle={() => { }} active={false} />
    const component = mount(element)
    expect(component.html()).toContain(`<nav class="menubar ">`)
  })

  it("select /blog on /", () => {
    window.history.pushState({}, "", "/")
    const element = <Menubar toggle={() => { }} active="true" />
    const component = mount(element)
    expect(component.html()).toContain(`<li class="menu-item selected"><a aria-current="page" class="" href="/">Blog</a></li>`)
  })

  it("select /blog on /blog/this-is-a-blog-post", () => {
    window.history.pushState({}, "", "/blog/this-is-a-blog-post")
    const element = <Menubar toggle={() => { }} active="true" />
    const component = mount(element)
    expect(component.html()).toContain(`<li class="menu-item selected"><a aria-current="page" class="" href="/">Blog</a></li>`)
  })

  it("select /tags on /tags/technical", () => {
    window.history.pushState({}, "", "/tags/technical")
    const element = <Menubar toggle={() => { }} active="true" />
    const component = mount(element)
    expect(component.html()).toContain(`<li class="menu-item selected"><a href="/tags">Tags</a></li>`)
  })

  it("select /privacy-policy", () => {
    window.history.pushState({}, "", "/privacy-policy")
    const element = <Menubar toggle={() => { }} active="true" />
    const component = mount(element)
    expect(component.html()).toContain(`<li class="menu-item selected"><a href="/privacy-policy">Privacy Policy</a></li>`)
  })

  it("select nothing when route does not exist", () => {
    window.history.pushState({}, "", "/tags2/technical")
    const element = <Menubar toggle={() => { }} active="true" />
    const component = mount(element)
    expect(component.html().indexOf(`<li class="menu-item selected">`)).toEqual(-1)
  })

  it("display a link to a page", () => {
    const element = <Menubar toggle={() => { }} active="true" />
    const component = mount(element)
    expect(component.html()).toContain(`<a href="/privacy-policy">Privacy Policy</a>`)
  })

  it("should toggle menubar", () => {
    let arbitraryNumber = 0
    const element = <Menubar toggle={() => arbitraryNumber =(arbitraryNumber === 0) ? 1 : 0} active="true" />
    const component = renderer.create(element)
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()

    tree.props.onClick()
    tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    expect(arbitraryNumber).toEqual(1)

    tree.props.onClick();
    tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    expect(arbitraryNumber).toEqual(0)
  })
})