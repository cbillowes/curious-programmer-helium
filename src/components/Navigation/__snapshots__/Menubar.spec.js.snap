// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`a menu bar component should render 1`] = `
<nav
  className="menubar active"
  onClick={[Function]}
>
  <ul>
    <li
      className="menu-item selected"
    >
      <a
        aria-current="page"
        className=""
        href="/"
        onClick={[Function]}
        onMouseEnter={[Function]}
        style={Object {}}
      >
        Blog
      </a>
    </li>
    <li
      className="menu-item "
    >
      <a
        href="/about"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        About
      </a>
    </li>
    <li
      className="menu-item "
    >
      <a
        href="/tags"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        Tags
      </a>
    </li>
    <li
      className="menu-item "
    >
      <a
        href="/credits"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        Credits
      </a>
    </li>
    <li
      className="menu-item "
    >
      <a
        href="/privacy-policy"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        Privacy Policy
      </a>
    </li>
    <li
      className="menu-item "
    >
      <a
        href="/license"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        License
      </a>
    </li>
  </ul>
</nav>
`;

exports[`a menu bar component should toggle menubar 1`] = `
<nav
  className="menubar active"
  onClick={[Function]}
>
  <ul>
    <li
      className="menu-item "
    >
      <a
        aria-current="page"
        className=""
        href="/"
        onClick={[Function]}
        onMouseEnter={[Function]}
        style={Object {}}
      >
        Blog
      </a>
    </li>
    <li
      className="menu-item "
    >
      <a
        href="/about"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        About
      </a>
    </li>
    <li
      className="menu-item "
    >
      <a
        href="/tags"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        Tags
      </a>
    </li>
    <li
      className="menu-item "
    >
      <a
        href="/credits"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        Credits
      </a>
    </li>
    <li
      className="menu-item "
    >
      <a
        href="/privacy-policy"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        Privacy Policy
      </a>
    </li>
    <li
      className="menu-item "
    >
      <a
        href="/license"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        License
      </a>
    </li>
  </ul>
</nav>
`;

exports[`a menu bar component should toggle menubar 2`] = `
<nav
  className="menubar active"
  onClick={[Function]}
>
  <ul>
    <li
      className="menu-item "
    >
      <a
        aria-current="page"
        className=""
        href="/"
        onClick={[Function]}
        onMouseEnter={[Function]}
        style={Object {}}
      >
        Blog
      </a>
    </li>
    <li
      className="menu-item "
    >
      <a
        href="/about"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        About
      </a>
    </li>
    <li
      className="menu-item "
    >
      <a
        href="/tags"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        Tags
      </a>
    </li>
    <li
      className="menu-item "
    >
      <a
        href="/credits"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        Credits
      </a>
    </li>
    <li
      className="menu-item "
    >
      <a
        href="/privacy-policy"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        Privacy Policy
      </a>
    </li>
    <li
      className="menu-item "
    >
      <a
        href="/license"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        License
      </a>
    </li>
  </ul>
</nav>
`;

exports[`a menu bar component should toggle menubar 3`] = `
<nav
  className="menubar active"
  onClick={[Function]}
>
  <ul>
    <li
      className="menu-item "
    >
      <a
        aria-current="page"
        className=""
        href="/"
        onClick={[Function]}
        onMouseEnter={[Function]}
        style={Object {}}
      >
        Blog
      </a>
    </li>
    <li
      className="menu-item "
    >
      <a
        href="/about"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        About
      </a>
    </li>
    <li
      className="menu-item "
    >
      <a
        href="/tags"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        Tags
      </a>
    </li>
    <li
      className="menu-item "
    >
      <a
        href="/credits"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        Credits
      </a>
    </li>
    <li
      className="menu-item "
    >
      <a
        href="/privacy-policy"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        Privacy Policy
      </a>
    </li>
    <li
      className="menu-item "
    >
      <a
        href="/license"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        License
      </a>
    </li>
  </ul>
</nav>
`;
